### ISSUE TITLE
It should be simple to walk a query tree without resolving it, but it isnt

### ISSUE THREAD

**Issue by lanwin** [graphql-dotnet/graphql-dotnet#2164]
My scenario is the following. One part of our graph is used for building aggregations. These are directly mapping to the Elasticsearch aggregation engine where you can build complex aggregations with sub aggreations multiple levels deep. 

It looks somewhat like this:
```graphql
{ 
  aggreagtions {
    timeSeries {
       topTerms {         
         topTerms {
            sum
            terms
          }
       }
    }
  }
}
```

The key part here is, that on the aggregation field we need to build the entire query...execute it and then resolve the results down the graph. 

My first attempt was to use the IDataLoaderResult, resolve down the graph and execute the query if all DataLoaderResults are awaiting. But that entire thing went pretty complex and hard to follow. 

So I switched to another solution. On the aggregation field I pre walk down the graph and walk the graph, build the query, execute it and then only resolve the executed results. That worked well end is a lot simple to follow.

But it was pretty harder to do so with Graphql-Dotnet cause everything is currently build around resolving the graph. I came up with this class to assist me while walking the tree:

```csharp
public class FieldContext
{
    FieldContext()
    {
    }

    public FieldContext(IResolveFieldContext context)
    {
        FieldAst = context.FieldAst;
        FiledDefinition = ExecutionHelper.GetFieldDefinition(context.Schema, context.ParentType, FieldAst);
        ExecutionContext = new ExecutionContext {Fragments = context.Fragments, Schema = context.Schema, Variables = context.Variables};
    }

    public string Name => FieldAst.Name;
    public string AliasOrName => FieldAst.GetAliasOrFieldName();

    public ExecutionContext ExecutionContext { get; private set; }
    public FieldType FiledDefinition { get; private set; }
    public Field FieldAst { get; private set; }

    public IEnumerable<FieldContext> GetFields()
    {
        var objectGraph = ResolveType(FiledDefinition.ResolvedType) as IObjectGraphType;
        if(objectGraph == null) yield break;

        foreach(var field in ExecutionHelper.SubFieldsFor(ExecutionContext, objectGraph, FieldAst))
        {
            var fieldDefinition = ExecutionHelper.GetFieldDefinition(ExecutionContext.Schema, objectGraph, field.Value);
            yield return new FieldContext
            {
                FieldAst = field.Value,
                FiledDefinition = fieldDefinition,
                ExecutionContext = ExecutionContext,
            };
        }
    }

    public T GetArgument<T>(string name)
    {
        var field = FieldAst.Arguments?.FirstOrDefault(a => a.Name == name);
        var valueValue = field?.Value;

        var filedDefinition = FiledDefinition.Arguments.FirstOrDefault(x => x.Name == name);
        if(valueValue == null)
            return (T)filedDefinition?.DefaultValue;

        var value = valueValue switch
        {
            VariableReference reference => ExecutionContext.Variables.FirstOrDefault(v => v.Name == reference.Name)?.Value,
            EnumValue => StringUtils.ToCamelCase(valueValue.Value.ToString()),
            _ => valueValue?.Value
        };

        return GraphQlExtensions.ArgumentDeserialized<T>(value);
    }

    IGraphType ResolveType(IGraphType type) =>
        type switch
        {
            NonNullGraphType nullType => ResolveType(nullType.ResolvedType),
            ListGraphType listType => ResolveType(listType.ResolvedType),
            _ => type
        };
}
```

But I didnt like it very much and I think a solution for this should be build in to Graphql-Dotnet. 

I would offer my help, but for now I have no good idea how a good general solution could look like. 

The hardest part is it to join FieldAst with field default value and query values. Maybe it would help if the IResolveFieldContext would have a function to get the IResolveFieldContext for all sub fields. Maybe there could be a IFieldContext witch inherits IResolveFieldContext from and a function to get the FieldContext for all child nodes or something like that.

**Comment by Shane32** [comment]
Relates to #354 : Allow resolvers to easily see what child fields are required.

I'm thinking:
- We need something implemented as an extension method on `IResolveFieldContext`.
- It should be capable of walking the entire descendant tree (or another method provided for that purpose).

We could implement this as a method that returns a list of children nodes, or a 'walker' which executes `Action<T>` methods for child nodes.  Or both.

Also if we implement #2016 _Precompile an execution tree..._ then it might be very easy (non-cpu intensive) to return these results.

I think what we need to figure out is the signatures for the methods.  The implementation will be the easy part.

I have a similar but different use case myself: I wrote code to build a LINQ `Select` expression based on the the fields and descendant fields requested.

Any ideas offhand, @sungam3r ?

**Comment by sungam3r** [comment]
`ExecutionStrategy` is a _client_ for `GraphQL.Language.AST.Document` APIs. This class represents a document tree that can be traversed for any purpose. Yes, we can provide some classes for such traversal. In general, the already existing classes for validation are quite suitable if you generalize them so that they work not only with `ValidationContext`.

**Comment by Shane32** [comment]
The problem is that the validation APIs don't traverse the tree as seen in the result set.  For example, it doesn't follow a fragment.  I'm not sure exactly what the answer is here, but using the validation walker doesn't really help without a few changes.

**Comment by lanwin** [comment]
What about making a FieldContext base class witch inherits ResolveFieldContext from. Mainly its everything except resolved stuff. So all extensions methods like GetArgument could be based on FieldContext instead of ResolveFieldContext. All whats needed then is a GetSubFields extension method that returns a list of FieldContext for every field like my class above does.

**Comment by sungam3r** [comment]
It seems to me that this issue will require a lot of changes and I am not yet ready to think about it on the eve of the v4 release.

**Comment by jannickj** [comment]
So happy someone else solved this problem for :heart: , I am also running my own query builder that needs the entire graph up front to do proper planning, I don't understand why it has been made this hard to get the resolved graph. It seems completely insane to me that the only way to use this library is that each field does its own call to the database...

**Comment by Shane32** [comment]
I also have a similar need and use case. I wrote my own code to walk the AST child nodes, but the code is incomplete. 

The problem is that the child node tree is not easy to traverse and is tightly integrated to the GraphQL.NET pipeline. There’s interfaces, fragments, unions, object graphs and leaf nodes, and fields can be aliased, resulting in two nodes that point to the same field. Once you have a field it needs to pull arguments from inline or query arguments or variables. 

A better way of walking an unexecuted AST tree which handles these nuances would certainly be useful.

**Comment by Shane32** [comment]
I think some progress was made with refactoring bow the SubFields property of IResolveFieldContext but I can’t remember if it’s useful past the first level of children.

**Comment by jannickj** [comment]
@Shane32 I just tried to use the solution above however it's very old and lots of the functions have been removed. Do you have the code for your graph traverser?

**Comment by jannickj** [comment]
For some reason the functions that allowed to do these things was removed in v4 thinking of `ExecutionHelper.SubFieldsFor`

**Comment by jannickj** [comment]
@Shane32 I don't understand your comment, I've just been looking through the PR history and you've been an instrumental force in making this both harder and more hacky todo, all without providing any alternative.

**Comment by jannickj** [comment]
I've created a new and updated version, some features of it:

* No reflection (Exposed Functions only)
* Resolves Fragments Correctly
* Resolves Arguments Correctly
* Resolves (skip/include) directives correctly
* Added functionality for auto casting Enumerable Arguments
* Uses the nullable feature of csharp

```csharp
public record FieldContext(
	ExecutionContext ExecutionContext,
	FieldType FiledDefinition,
	Field FieldAst
)
{
	public string Name => FieldAst.Name;
	public string AliasOrName => FieldAst.Alias ?? FieldAst.Name;

	public static FieldContext FromContext(IResolveFieldContext context)
	{
		return new FieldContext(
			FieldAst: context.FieldAst,
			FiledDefinition: GetFieldDefinition(context.Schema, context.ParentType, context.FieldAst)!,
			ExecutionContext: new ExecutionContext
			{
				Document = context.Document,
				Schema = context.Schema,
				Variables = context.Variables
			}
		);

	}


	public IEnumerable<FieldContext> Fields
	{
		get
		{
			if (FiledDefinition.ResolvedType is null) yield break;
			if (ResolveType(FiledDefinition.ResolvedType) is not IObjectGraphType objectGraph) yield break;

			var fields = GetSubFields(ExecutionContext, objectGraph, FieldAst);
			if (fields is null) yield break;
			foreach (var field in fields)
			{
				var fieldDefinition = GetFieldDefinition(ExecutionContext.Schema, objectGraph, field.Value);
				if (fieldDefinition is null) continue;
				yield return new FieldContext
				(
					FieldAst: field.Value,
					FiledDefinition: fieldDefinition,
					ExecutionContext: ExecutionContext
				);
			}
		}
	}

	public object? GetArgumentObject(string name)
	{
		var field = FieldAst.Arguments?.FirstOrDefault(a => a.Name == name);
		var valueValue = field?.Value;

		var filedDefinition = FiledDefinition.Arguments?.FirstOrDefault(x => x.Name == name);
		if (filedDefinition is null) return null;
		var argVal = ExecutionHelper.CoerceValue(
			filedDefinition.ResolvedType!, valueValue, ExecutionContext.Variables, filedDefinition.DefaultValue);
		return argVal.Value;
	}
	public IEnumerable<T> GetArgumentList<T>(string name)
	{
		var obj = GetArgumentObject(name);
		if (obj is null) yield break;
		var objs = (IEnumerable)obj;
		foreach (var o in objs)
		{
			yield return (T)o;
		}
	}

	IGraphType ResolveType(IGraphType type) =>
		type switch
		{
			NonNullGraphType nullType => ResolveType(nullType.ResolvedType!),
			ListGraphType listType => ResolveType(listType.ResolvedType!),
			_ => type
		};

	bool IsList(IGraphType type) =>
		type switch
		{
			NonNullGraphType nullType => IsList(nullType.ResolvedType!),
			ListGraphType => true,
			_ => false
		};


	private class GetAroundProtectedFunctions : ExecutionStrategy
	{
		protected override Task ExecuteNodeTreeAsync(ExecutionContext context, ObjectExecutionNode rootNode)
		{
			throw new System.NotImplementedException();
		}

		public Dictionary<string, Field> PublicCollectFieldsFrom(ExecutionContext context, IGraphType specificType, SelectionSet selectionSet, Dictionary<string, Field>? fields)
		{
			return base.CollectFieldsFrom(context, specificType, selectionSet, fields);
		}

		public FieldType? PublicGetFieldDefinition(ISchema schema, IObjectGraphType parentType, Field field)
		{
			return base.GetFieldDefinition(schema, parentType, field);
		}
	}

	private static readonly GetAroundProtectedFunctions _GetAroundProtectedFunctions = new();

	private static Dictionary<string, Field>? GetSubFields(ExecutionContext context, IGraphType graphType, Field field)
	{
		if (field == null || !(field!.SelectionSet?.Selections?.Count > 0))
		{
			return null;
		}
		return _GetAroundProtectedFunctions.PublicCollectFieldsFrom(
			context, graphType, field!.SelectionSet!, null);
	}

	private static FieldType? GetFieldDefinition(ISchema schema, IObjectGraphType parentType, Field field)
	{
		return _GetAroundProtectedFunctions.PublicGetFieldDefinition(schema, parentType, field);
	}
}

public static class FieldContextStruct
{
	public static T? GetArgument<T>(this FieldContext con, string name) where T : struct
	{
		return (T?)con.GetArgumentObject(name);
	}
}

public static class FieldContextClass
{
	public static T? GetArgument<T>(this FieldContext con, string name) where T : class
	{
		return (T?)con.GetArgumentObject(name);
	}
}
```

**Comment by Shane32** [comment]
> @Shane32 I just tried to use the solution above however it's very old and lots of the functions have been removed. Do you have the code for your graph traverser?

It's not public, and it's not worth looking at.  Again, it doesn't handle fragments or anything.  I can get away with it because it is not a public API.

**Comment by jannickj** [comment]
> It's not public, and it's not worth looking at. Again, it doesn't handle fragments or anything. I can get away with it because it is not a public API.

@Shane32 no worries I've posted a solution that handles all that.

**Comment by Shane32** [comment]
> For some reason the functions that allowed to do these things was removed in v4 thinking of `ExecutionHelper.SubFieldsFor`

It was removed from ExecutionHelper by necessity and moved into the execution strategy as a public method.  This consolidated the code of traversing the AST nodes -- there was some other reason too, but I forgot now.

**Comment by Shane32** [comment]
> @Shane32 no worries I've posted a solution that handles all that.

I appreciate it.  Perhaps we can consider integrating it as a new feature for v5 somehow.  I have not really looked at your code yet, but I will.

**Comment by lanwin** [comment]
> For any poor soul in the same situation as I was in: I've created a new and updated version of @lanwin 's code such that it works in v4.6, the features of this version is:
> 
>     * No reflection (Exposed Functions only)
> 
>     * Resolves Fragments Correctly
> 
>     * Resolves Arguments Correctly
> 
>     * Resolves (skip/include) directives correctly
> 
>     * Added functionality for auto casting Enumerable Arguments
> 
>     * Uses the nullable feature of csharp
> 
> 
> ```cs
> public record FieldContext(
> 	ExecutionContext ExecutionContext,
> 	FieldType FiledDefinition,
> 	Field FieldAst
> )
> {
> 	public string Name => FieldAst.Name;
> 	public string AliasOrName => FieldAst.Alias ?? FieldAst.Name;
> 
> 	public static FieldContext FromContext(IResolveFieldContext context)
> 	{
> 		return new FieldContext(
> 			FieldAst: context.FieldAst,
> 			FiledDefinition: GetFieldDefinition(context.Schema, context.ParentType, context.FieldAst)!,
> 			ExecutionContext: new ExecutionContext
> 			{
> 				Document = context.Document,
> 				Schema = context.Schema,
> 				Variables = context.Variables
> 			}
> 		);
> 
> 	}
> 
> 
> 	public IEnumerable<FieldContext> Fields
> 	{
> 		get
> 		{
> 			if (FiledDefinition.ResolvedType is null) yield break;
> 			if (ResolveType(FiledDefinition.ResolvedType) is not IObjectGraphType objectGraph) yield break;
> 
> 			var fields = GetSubFields(ExecutionContext, objectGraph, FieldAst);
> 			if (fields is null) yield break;
> 			foreach (var field in fields)
> 			{
> 				var fieldDefinition = GetFieldDefinition(ExecutionContext.Schema, objectGraph, field.Value);
> 				if (fieldDefinition is null) continue;
> 				yield return new FieldContext
> 				(
> 					FieldAst: field.Value,
> 					FiledDefinition: fieldDefinition,
> 					ExecutionContext: ExecutionContext
> 				);
> 			}
> 		}
> 	}
> 
> 	public object? GetArgumentObject(string name)
> 	{
> 		var field = FieldAst.Arguments?.FirstOrDefault(a => a.Name == name);
> 		var valueValue = field?.Value;
> 
> 		var filedDefinition = FiledDefinition.Arguments?.FirstOrDefault(x => x.Name == name);
> 		if (filedDefinition is null) return null;
> 		var argVal = ExecutionHelper.CoerceValue(
> 			filedDefinition.ResolvedType!, valueValue, ExecutionContext.Variables, filedDefinition.DefaultValue);
> 		return argVal.Value;
> 	}
> 	public IEnumerable<T> GetArgumentList<T>(string name)
> 	{
> 		var obj = GetArgumentObject(name);
> 		if (obj is null) yield break;
> 		var objs = (IEnumerable)obj;
> 		foreach (var o in objs)
> 		{
> 			yield return (T)o;
> 		}
> 	}
> 
> 	IGraphType ResolveType(IGraphType type) =>
> 		type switch
> 		{
> 			NonNullGraphType nullType => ResolveType(nullType.ResolvedType!),
> 			ListGraphType listType => ResolveType(listType.ResolvedType!),
> 			_ => type
> 		};
> 
> 	private class GetAroundProtectedFunctions : ExecutionStrategy
> 	{
> 		protected override Task ExecuteNodeTreeAsync(ExecutionContext context, ObjectExecutionNode rootNode)
> 		{
> 			throw new System.NotImplementedException();
> 		}
> 
> 		public Dictionary<string, Field> PublicCollectFieldsFrom(ExecutionContext context, IGraphType specificType, SelectionSet selectionSet, Dictionary<string, Field>? fields)
> 		{
> 			return base.CollectFieldsFrom(context, specificType, selectionSet, fields);
> 		}
> 
> 		public FieldType? PublicGetFieldDefinition(ISchema schema, IObjectGraphType parentType, Field field)
> 		{
> 			return base.GetFieldDefinition(schema, parentType, field);
> 		}
> 	}
> 
> 	private static readonly GetAroundProtectedFunctions _GetAroundProtectedFunctions = new();
> 
> 	private static Dictionary<string, Field>? GetSubFields(ExecutionContext context, IGraphType graphType, Field field)
> 	{
> 		if (field == null || !(field!.SelectionSet?.Selections?.Count > 0))
> 		{
> 			return null;
> 		}
> 		return _GetAroundProtectedFunctions.PublicCollectFieldsFrom(
> 			context, graphType, field!.SelectionSet!, null);
> 	}
> 
> 	private static FieldType? GetFieldDefinition(ISchema schema, IObjectGraphType parentType, Field field)
> 	{
> 		return _GetAroundProtectedFunctions.PublicGetFieldDefinition(schema, parentType, field);
> 	}
> }
> 
> public static class FieldContextStruct
> {
> 	public static T? GetArgument<T>(this FieldContext con, string name) where T : struct
> 	{
> 		return (T?)con.GetArgumentObject(name);
> 	}
> }
> 
> public static class FieldContextClass
> {
> 	public static T? GetArgument<T>(this FieldContext con, string name) where T : class
> 	{
> 		return (T?)con.GetArgumentObject(name);
> 	}
> }
> ```

Thank you for migrating my code.

**Comment by sungam3r** [comment]
@lanwin I'm not sure (too long ro read all comments) but it seems that this task may be solved by new visitors API from [parser project](https://github.com/graphql-dotnet/parser/tree/master/src/GraphQLParser/Visitors). v8 is not published yet. Visitor works with some `ASTNode` and traverses this node down the tree calling `VisitXXX`. You override required methods and gather some info. We use that API in complexity analizer for example. Also see [readme](https://github.com/graphql-dotnet/parser#3-astvisitor) .

**Comment by Shane32** [comment]
I don’t think that will help. The problem is trying to walk the list of nodes that will execute. Not the AST. It needs to follow fragments and maybe even parse variables and literals. Basically, return a “resolve field context” for each child field

**Comment by sungam3r** [comment]
>  that will execute. Not the AST

😞

**Comment by lanwin** [comment]
> > that will execute. Not the AST
> 
> 😞

@Shane32 is right.

**Comment by sungam3r** [comment]
Note - there may be a problem with fields that return interfaces.

**Comment by Shane32** [comment]
Since the type isn't known until it is resolved, it isn't known with certainty which fields will execute.  We could also this is true of any field -- if the parent returns null, the child will not execute.

Most likely we could return them all, I expect, with the understanding that not all fields walked may execute, depending on the values returned in parent nodes.

**Comment by lanwin** [comment]
> For any poor soul in the same situation as I was in: I've created a new and updated version of @lanwin 's code such that it works in v4.6, the features of this version is:
> 
>     * No reflection (Exposed Functions only)
> 
>     * Resolves Fragments Correctly
> 
>     * Resolves Arguments Correctly
> 
>     * Resolves (skip/include) directives correctly
> 
>     * Added functionality for auto casting Enumerable Arguments
> 
>     * Uses the nullable feature of csharp
> 
> 
> ```cs
> public record FieldContext(
> 	ExecutionContext ExecutionContext,
> 	FieldType FiledDefinition,
> 	Field FieldAst
> )
> {
> 	public string Name => FieldAst.Name;
> 	public string AliasOrName => FieldAst.Alias ?? FieldAst.Name;
> 
> 	public static FieldContext FromContext(IResolveFieldContext context)
> 	{
> 		return new FieldContext(
> 			FieldAst: context.FieldAst,
> 			FiledDefinition: GetFieldDefinition(context.Schema, context.ParentType, context.FieldAst)!,
> 			ExecutionContext: new ExecutionContext
> 			{
> 				Document = context.Document,
> 				Schema = context.Schema,
> 				Variables = context.Variables
> 			}
> 		);
> 
> 	}
> 
> 
> 	public IEnumerable<FieldContext> Fields
> 	{
> 		get
> 		{
> 			if (FiledDefinition.ResolvedType is null) yield break;
> 			if (ResolveType(FiledDefinition.ResolvedType) is not IObjectGraphType objectGraph) yield break;
> 
> 			var fields = GetSubFields(ExecutionContext, objectGraph, FieldAst);
> 			if (fields is null) yield break;
> 			foreach (var field in fields)
> 			{
> 				var fieldDefinition = GetFieldDefinition(ExecutionContext.Schema, objectGraph, field.Value);
> 				if (fieldDefinition is null) continue;
> 				yield return new FieldContext
> 				(
> 					FieldAst: field.Value,
> 					FiledDefinition: fieldDefinition,
> 					ExecutionContext: ExecutionContext
> 				);
> 			}
> 		}
> 	}
> 
> 	public object? GetArgumentObject(string name)
> 	{
> 		var field = FieldAst.Arguments?.FirstOrDefault(a => a.Name == name);
> 		var valueValue = field?.Value;
> 
> 		var filedDefinition = FiledDefinition.Arguments?.FirstOrDefault(x => x.Name == name);
> 		if (filedDefinition is null) return null;
> 		var argVal = ExecutionHelper.CoerceValue(
> 			filedDefinition.ResolvedType!, valueValue, ExecutionContext.Variables, filedDefinition.DefaultValue);
> 		return argVal.Value;
> 	}
> 	public IEnumerable<T> GetArgumentList<T>(string name)
> 	{
> 		var obj = GetArgumentObject(name);
> 		if (obj is null) yield break;
> 		var objs = (IEnumerable)obj;
> 		foreach (var o in objs)
> 		{
> 			yield return (T)o;
> 		}
> 	}
> 
> 	IGraphType ResolveType(IGraphType type) =>
> 		type switch
> 		{
> 			NonNullGraphType nullType => ResolveType(nullType.ResolvedType!),
> 			ListGraphType listType => ResolveType(listType.ResolvedType!),
> 			_ => type
> 		};
> 
> 	private class GetAroundProtectedFunctions : ExecutionStrategy
> 	{
> 		protected override Task ExecuteNodeTreeAsync(ExecutionContext context, ObjectExecutionNode rootNode)
> 		{
> 			throw new System.NotImplementedException();
> 		}
> 
> 		public Dictionary<string, Field> PublicCollectFieldsFrom(ExecutionContext context, IGraphType specificType, SelectionSet selectionSet, Dictionary<string, Field>? fields)
> 		{
> 			return base.CollectFieldsFrom(context, specificType, selectionSet, fields);
> 		}
> 
> 		public FieldType? PublicGetFieldDefinition(ISchema schema, IObjectGraphType parentType, Field field)
> 		{
> 			return base.GetFieldDefinition(schema, parentType, field);
> 		}
> 	}
> 
> 	private static readonly GetAroundProtectedFunctions _GetAroundProtectedFunctions = new();
> 
> 	private static Dictionary<string, Field>? GetSubFields(ExecutionContext context, IGraphType graphType, Field field)
> 	{
> 		if (field == null || !(field!.SelectionSet?.Selections?.Count > 0))
> 		{
> 			return null;
> 		}
> 		return _GetAroundProtectedFunctions.PublicCollectFieldsFrom(
> 			context, graphType, field!.SelectionSet!, null);
> 	}
> 
> 	private static FieldType? GetFieldDefinition(ISchema schema, IObjectGraphType parentType, Field field)
> 	{
> 		return _GetAroundProtectedFunctions.PublicGetFieldDefinition(schema, parentType, field);
> 	}
> }
> 
> public static class FieldContextStruct
> {
> 	public static T? GetArgument<T>(this FieldContext con, string name) where T : struct
> 	{
> 		return (T?)con.GetArgumentObject(name);
> 	}
> }
> 
> public static class FieldContextClass
> {
> 	public static T? GetArgument<T>(this FieldContext con, string name) where T : class
> 	{
> 		return (T?)con.GetArgumentObject(name);
> 	}
> }
> ```

@jannickj there is one missing pice in the code you have updated. GetArgument needs to call the ValueConverter.

```csharp
public static class FieldContextStruct
{
   public static T? GetArgument<T>(this FieldContext con, string name)
   where T : struct => ValueConverter.ConvertTo<T?>(con.GetArgumentObject(name));
}
public static class FieldContextClass
{
   public static T? GetArgument<T>(this FieldContext con, string name)
         where T : class => ValueConverter.ConvertTo<T?>(con.GetArgumentObject(name));
}
```
